src/MarginTrading.NotificationGenerator.Services/TradingReportService.cs

```diff
@@ -0,0 +1,206 @@
+                .ToList();
+            var openPositions = accountHistoryAggregate.OpenPositions.Where(x => accountIds.Contains(x.AccountId))
+                .Select(x => _convertService.Convert<OrderHistoryContract, OrderHistory>(x))
+                .SetClientId(accountClients)
+                .ToList();
+            var pendingPositions = (await pendingPositionsTask).Where(x => accountIds.Contains(x.AccountId))
+                .Select(x => _convertService.Convert<OrderContract, OrderHistory>(x))
+                .ToList();
+            var accountTransactions = accountHistoryAggregate.Account.Where(x => accountIds.Contains(x.AccountId))
+                .Select(x => _convertService.Convert<AccountHistoryContract, AccountHistory>(x))
+                .ToList();
+
+            //prepare notification models
+            var notifications = clientIds.Select(x =>
+                PrepareNotification(reportMonth, from, to, x, closedTrades, openPositions, pendingPositions,
+                    accounts, accountTransactions)).ToList();
+
+            //retrieve emails
+            var emails = (await _clientAccountClient.GetClientsByIdsAsync(clientIds))
+                .ToDictionary(x => x.Id, x => x.Email);
+
+            await SendNotifications(notifications, emails);
+        }
+
+        private async Task SendNotifications(IEnumerable<MonthlyTradingNotification> notifications, 
+            IReadOnlyDictionary<string, string> emails)
+        {
+            var failedNotifications = new Dictionary<MonthlyTradingNotification, Exception>();
+            Exception anyException = null;
+            foreach (var notification in notifications)
+            {
+                try
+                {
+                    await _emailService.PrepareAndSendEmailAsync(emails[notification.ClientId],
+                        $"Margin Trading - Monthly trading report for {notification.CurrentMonth}",
+                        "MonthlyTradingReport",
+                        notification);
+
+                    await _log.WriteInfoAsync(nameof(NotificationGenerator), nameof(TradingReportService),
+                        nameof(PerformReporting), notification.GetLogData(), _systemClock.UtcNow.DateTime);
+                }
+                catch (Exception ex)
+                {
+                    anyException = ex;
+                    failedNotifications.Add(notification, ex);
+                }
+            }
+
+            if (failedNotifications.Any())
+            {
+                //TODO handle failed notifications
+                await _log.WriteWarningAsync(nameof(NotificationGenerator), nameof(TradingReportService),
+                    nameof(PerformReporting), $"{failedNotifications.Count} notifications failed to be send.",
+                    anyException, _systemClock.UtcNow.DateTime);
+            }
+        }
+
+        private static MonthlyTradingNotification PrepareNotification(int reportMonth, 
```

Comment:
Very strange manipulations with collections... Let's discuss it.

Reply:
Might be optimised for sure. But it's ok for the particular case.